### PR TITLE
Refactor load in checkpoint connector

### PR DESCRIPTION
**Pull request by tarepan**
## What does this PR do?
Refactor checkpoint load in CheckpointConnector.  
Only refactoring for code-health, no functional changes.  
Two main refactored points:  

- align `restore` (normal load) & `hpc_load` (hpc load)
- common hpc file check/selection

Commits are well-separated based on refactored parts.  
If changes are no apparent in "Files changed" diff, please check "Commits" diff.  

## Before submitting
- [n/a] refactoring <= Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [n/a] Did you make sure to update the documentation with your changes?
- [n/a] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In in short, see following bullet-list:  

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Make sure you had fun coding 👍

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -117,14 +110,14 @@ def restore_model_state(self, model: LightningModule, checkpoint) -> None:
         Restore model states from a 'PyTorch-Lightning checkpoint' dictionary object
         """
 
-        # give the datamodule a chance to load something
+        # restore datamodule states
         if self.trainer.datamodule is not None:
             self.trainer.datamodule.on_load_checkpoint(checkpoint)
 
-        # give model a chance to restore something
+        # hook for arbitrary processing before model restore
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:117]:
```suggestion
        # hook: give user access to checkpoint if needed. 
```

File: tests/base/develop_pipelines.py [hunk 2]
```diff
@@ -86,9 +86,13 @@ def run_model_test(trainer_options, model, on_gpu: bool = True, version=None, wi
             trainer.optimizers, trainer.lr_schedulers, trainer.optimizer_frequencies = \
                 trainer.init_optimizers(pretrained_model)
 
-        # test HPC loading / saving
+        # test HPC saving
         trainer.checkpoint_connector.hpc_save(save_dir, logger)
-        trainer.checkpoint_connector.hpc_load(save_dir, on_gpu=on_gpu)
+        # test HPC loading
+        max_suffix = trainer.checkpoint_connector.max_ckpt_in_folder(save_dir)
```
Comment (tchaton) [inline tests/base/develop_pipelines.py:92]: Could this be wrapped into a `get_max_checkpoint_path_from_folder` function ?

Comment (tarepan) [reply tests/base/develop_pipelines.py]: Refactored, thanks for review!

**Review by justusschock — APPROVED** [review]

**Comment by Borda** [comment]
@tarepan how is it going here, ready to co?
cc: @tchaton

**Comment by tarepan** [comment]
@Borda Yes, ready to go.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 3]
```diff
@@ -330,36 +307,52 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
...
-        # load training state (affects trainer only)
+        # restore training state
         self.restore_training_state(checkpoint)
 
-        # call model hook
+        # call hpc specific hook
         model.on_hpc_load(checkpoint)
 
-        log.info(f'restored hpc model from: {filepath}')
+    def max_ckpt_in_folder(self, dir_path: Union[str, Path], name_key: str = 'ckpt_') -> Union[None, int]:
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:334]:
```suggestion
    def max_ckpt_in_folder(self, dir_path: Union[str, Path], name_key: str = 'ckpt_') -> Optional[int]:
```

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:334]: you may have to import Optional

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 4]
```diff
@@ -330,36 +307,52 @@ def dump_checkpoint(self, weights_only: bool = False) -> dict:
...
+        """List up files in `dir_path` with name_key, then yield maximum suffix number.
+
+        Args:
+            dir_path: path of directory which may contain files which name include `name_key`
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:338]:
```suggestion
            dir_path: path of directory which may contain files whose name include `name_key`
```

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:338]: I'm sure "which" is not correct here. I think one has to use "whose".

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py (hunk 3 above)
Comment (tarepan) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: Committed.
